samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -56,53 +54,75 @@
 public class LocalApplicationRunner extends AbstractApplicationRunner {
 
   private static final Logger log = LoggerFactory.getLogger(LocalApplicationRunner.class);
-  private static final String LATCH_INIT = "init";
-  private static final long LATCH_TIMEOUT_MINUTES = 10; // 10 min timeout
+  // Latch id that's used for awaiting the init of application before creating the StreamProcessors
+  private static final String INIT_LATCH_ID = "init";
+  // Latch timeout is set to 10 min
+  private static final long LATCH_TIMEOUT_MINUTES = 10;
 
   private final String uid;
-  private final CoordinationUtils coordination;
-  private final List<StreamProcessor> processors = new ArrayList<>();
-  private final CountDownLatch latch = new CountDownLatch(1);
-  private final AtomicReference<Throwable> throwable = new AtomicReference<>();
-  private final ConcurrentHashSet<String> processorIds = new ConcurrentHashSet<>();
+  private final CoordinationUtils coordinationUtils;
+  private final Set<StreamProcessor> processors = ConcurrentHashMap.newKeySet();
+  private final CountDownLatch shutdownLatch = new CountDownLatch(1);
+  private final AtomicInteger numProcessorsToStart = new AtomicInteger();
 
   private ApplicationStatus appStatus = ApplicationStatus.New;
 
-  final class LocalStreamProcessorListener implements StreamProcessorLifecycleListener {
-    public final String processorId;
+  final class LocalStreamProcessorLifeCycleListener implements StreamProcessorLifecycleListener {
+    StreamProcessor processor;
 
-    public LocalStreamProcessorListener(String processorId) {
-      if (StringUtils.isEmpty(processorId)) {
-        throw new NullPointerException("processorId has to be defined in LocalStreamProcessorListener.");
-      }
-      this.processorId = processorId;
+    void setProcessor(StreamProcessor processor) {
+      this.processor = processor;
     }
 
     @Override
     public void onStart() {
-      processorIds.add(processorId);
+      if (numProcessorsToStart.decrementAndGet() == 0) {
+        appStatus = ApplicationStatus.Running;
+      }
     }
 
     @Override
     public void onShutdown() {
-      processorIds.remove(processorId);
-      if (processorIds.isEmpty()) {
-        latch.countDown();
+      processors.remove(processor);
+      processor = null;
```

Comment:
Is this required (here and in onFailure)? Seems like the reference will go away when the listener reference is removed from the set.